Compiler crash with Shapeless 2.10

Trying to build [Shapeless 2.10](https://github.com/milessabin/shapeless/tree/scalajs-2.10.x), the main module `core` compiles fine but the `examples` module crashes the compiler.

You can see the whole log here in Travis:
https://travis-ci.org/milessabin/shapeless/jobs/57126130

Crashy bit is:
```
[error]      while compiling: /home/golly/projects/contrib/shapeless/examples/src/main/scala/shapeless/examples/shows.scala
[error]         during phase: jscode
[error]      library version: version 2.10.5
[error]     compiler version: version 2.10.5
[error]   reconstructed args: -language:higherKinds -language:implicitConversions -deprecation -feature -classpath /home/golly/projects/contrib/shapeless/examples/target/scala-2.10/classes:/home/golly/projects/contrib/shapeless/core/target/scala-2.10/classes:/home/golly/.ivy2/cache/org.scala-js/scalajs-library_2.10/jars/scalajs-library_2.10-0.6.2.jar:/home/golly/.ivy2/cache/org.scalamacros/quasiquotes_2.10/jars/quasiquotes_2.10-2.0.0.jar:/home/golly/.ivy2/cache/org.scala-lang/scala-compiler/jars/scala-compiler-2.10.5.jar:/home/golly/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.10.5.jar -P:scalajs:mapSourceURI:file:/home/golly/projects/contrib/shapeless/->https://raw.githubusercontent.com/milessabin/shapeless/shapeless-2.2.0-RC3/ -Xfatal-warnings -Yfundep-materialization -Xplugin:/home/golly/.ivy2/cache/org.scala-js/scalajs-compiler_2.10.5/jars/scalajs-compiler_2.10.5-0.6.2.jar -Xplugin:/home/golly/.ivy2/cache/org.scalamacros/paradise_2.10.5/jars/paradise_2.10.5-2.0.1.jar -unchecked -bootclasspath /usr/lib/jvm/java-8-jdk/jre/lib/resources.jar:/usr/lib/jvm/java-8-jdk/jre/lib/rt.jar:/usr/lib/jvm/java-8-jdk/jre/lib/sunrsasign.jar:/usr/lib/jvm/java-8-jdk/jre/lib/jsse.jar:/usr/lib/jvm/java-8-jdk/jre/lib/jce.jar:/usr/lib/jvm/java-8-jdk/jre/lib/charsets.jar:/usr/lib/jvm/java-8-jdk/jre/lib/jfr.jar:/usr/lib/jvm/java-8-jdk/jre/classes:/home/golly/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.10.5.jar
[error] 
[error]   last tree to typer: TypeTree(class Long)
[error]               symbol: class Long in package scala (flags: final abstract)
[error]    symbol definition: final abstract class Long extends 
[error]                  tpe: Long
[error]        symbol owners: class Long -> package scala
[error]       context owners: class ZipperExamples$delayedInit$body -> package examples
[error] 
[error] == Enclosing template or block ==
[error] 
[error] Template( // val <local delayedInit$body>: <notype> in class ZipperExamples$delayedInit$body, tree.tpe=shapeless.examples.ZipperExamples$delayedInit$body
[error]   "scala.runtime.AbstractFunction0" // parents
[error]   ValDef(
[error]     private
[error]     "_"
[error]     <tpt>
[error]     <empty>
[error]   )
[error]   // 3 statements
[error]   ValDef( // private[this] val $outer: shapeless.examples.ZipperExamples.type in class ZipperExamples$delayedInit$body
[error]     private <local> <paramaccessor> <triedcooking>
[error]     "$outer"
[error]     <tpt> // tree.tpe=shapeless.examples.ZipperExamples.type
[error]     <empty>
[error]   )
[error]   DefDef( // final def apply(): Object in class ZipperExamples$delayedInit$body
[error]     <method> final
[error]     "apply"
[error]     []
[error]     List(Nil)
[error]     <tpt> // tree.tpe=Object
[error]     Block( // tree.tpe=runtime.BoxedUnit
[error]       // 32 statements
[error]       Apply( // private[object ZipperExamples] def dept_=(x$1: shapeless.examples.ZipperExamples$Dept): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."dept_$eq" // private[object ZipperExamples] def dept_=(x$1: shapeless.examples.ZipperExamples$Dept): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.examples.ZipperExamples$Dept)Unit
[error]         Apply( // def <init>(manager: shapeless.examples.ZipperExamples$Employee,employees: shapeless.HList): shapeless.examples.ZipperExamples$Dept in class ZipperExamples$Dept, tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           new shapeless.examples.ZipperExamples$Dept."<init>" // def <init>(manager: shapeless.examples.ZipperExamples$Employee,employees: shapeless.HList): shapeless.examples.ZipperExamples$Dept in class ZipperExamples$Dept, tree.tpe=(manager: shapeless.examples.ZipperExamples$Employee, employees: shapeless.HList)shapeless.examples.ZipperExamples$Dept
[error]           // 2 arguments
[error]           Apply( // def <init>(name: String,salary: Int): shapeless.examples.ZipperExamples$Employee in class ZipperExamples$Employee, tree.tpe=shapeless.examples.ZipperExamples$Employee
[error]             new shapeless.examples.ZipperExamples$Employee."<init>" // def <init>(name: String,salary: Int): shapeless.examples.ZipperExamples$Employee in class ZipperExamples$Employee, tree.tpe=(name: String, salary: Int)shapeless.examples.ZipperExamples$Employee
[error]             // 2 arguments
[error]             "Agamemnon"
[error]             5000
[error]           )
[error]           Block( // tree.tpe=shapeless.::
[error]             ValDef( // val x$3: shapeless.examples.ZipperExamples$Employee
[error]               <synthetic> <triedcooking>
[error]               "x$3"
[error]               <tpt> // tree.tpe=shapeless.examples.ZipperExamples$Employee
[error]               Apply( // def <init>(name: String,salary: Int): shapeless.examples.ZipperExamples$Employee in class ZipperExamples$Employee, tree.tpe=shapeless.examples.ZipperExamples$Employee
[error]                 new shapeless.examples.ZipperExamples$Employee."<init>" // def <init>(name: String,salary: Int): shapeless.examples.ZipperExamples$Employee in class ZipperExamples$Employee, tree.tpe=(name: String, salary: Int)shapeless.examples.ZipperExamples$Employee
[error]                 // 2 arguments
[error]                 "Menelaus"
[error]                 3000
[error]               )
[error]             )
[error]             Apply( // def ::(h: Object): shapeless.:: in class HListOps, tree.tpe=shapeless.::
[error]               shapeless.this.HList.hlistOps({
[error]   <synthetic> val x$2: shapeless.examples.ZipperExamples$Employee = new shapeless.examples.ZipperExamples$Employee("Achilles", 2000);
[error]   shapeless.this.HList.hlistOps({
[error]   <synthetic> val x$1: shapeless.examples.ZipperExamples$Employee = new shapeless.examples.ZipperExamples$Employee("Odysseus", 2000);
[error]   shapeless.HNil.::(x$1)
[error] }).::(x$2)
[error] })."$colon$colon" // def ::(h: Object): shapeless.:: in class HListOps, tree.tpe=(h: Object)shapeless.::
[error]               "x$3" // val x$3: shapeless.examples.ZipperExamples$Employee, tree.tpe=shapeless.examples.ZipperExamples$Employee
[error]             )
[error]           )
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def z_=(x$1: shapeless.Zipper): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."z_$eq" // private[object ZipperExamples] def z_=(x$1: shapeless.Zipper): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.Zipper)Unit
[error]         Apply( // def toZipper(gen: shapeless.Generic): shapeless.Zipper in class GenericZipperOps, tree.tpe=shapeless.Zipper
[error]           shapeless.syntax.zipper.toZipper(ZipperExamples$delayedInit$body.this.$outer.dept(), {
[error]   new anonymous class anon$1()
[error] })."toZipper" // def toZipper(gen: shapeless.Generic): shapeless.Zipper in class GenericZipperOps, tree.tpe=(gen: shapeless.Generic)shapeless.Zipper
[error]           Block( // tree.tpe=shapeless.examples.1
[error]             {}
[error]             Apply( // def <init>(): shapeless.examples.1 in class 1, tree.tpe=shapeless.examples.1
[error]               new shapeless.examples.1."<init>" // def <init>(): shapeless.examples.1 in class 1, tree.tpe=()shapeless.examples.1
[error]               Nil
[error]             )
[error]           )
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def g1_=(x$1: shapeless.examples.ZipperExamples$Dept): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."g1_$eq" // private[object ZipperExamples] def g1_=(x$1: shapeless.examples.ZipperExamples$Dept): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.examples.ZipperExamples$Dept)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()shapeless.examples.ZipperExamples$Dept
[error]             ZipperExamples$delayedInit$body.this.$outer.z().reify(shapeless.ops.zipper$Reify.genericReify({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // def typed(t: Function0): Unit in package test, tree.tpe=Unit
[error]         "shapeless"."test"."package"."typed" // def typed(t: Function0): Unit in package test, tree.tpe=(t: Function0)Unit
[error]         Block( // tree.tpe=Function0
[error]           {}
[error]           Typed( // tree.tpe=Function0
[error]             Apply( // def <init>(): shapeless.examples.anonfun$2, tree.tpe=shapeless.examples.anonfun$2
[error]               new anonymous class anonfun$2."<init>" // def <init>(): shapeless.examples.anonfun$2, tree.tpe=()shapeless.examples.anonfun$2
[error]               Nil
[error]             )
[error]             <tpt> // tree.tpe=Function0
[error]           )
[error]         )
[error]       )
[error]       Apply( // def println(x: Object): Unit in object Predef, tree.tpe=Unit
[error]         scala.this."Predef"."println" // def println(x: Object): Unit in object Predef, tree.tpe=(x: Object)Unit
[error]         Apply( // val g1(): shapeless.examples.ZipperExamples$Dept in object ZipperExamples, tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           ZipperExamples$delayedInit$body.this."$outer"."g1" // val g1(): shapeless.examples.ZipperExamples$Dept in object ZipperExamples, tree.tpe=()shapeless.examples.ZipperExamples$Dept
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def g2_=(x$1: shapeless.::): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."g2_$eq" // private[object ZipperExamples] def g2_=(x$1: shapeless.::): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.::)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=shapeless.::
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()shapeless.::
[error]             ZipperExamples$delayedInit$body.this.$outer.z().right(shapeless.ops.zipper$Right.right()).$asInstanceOf[shapeless.Zipper]().get(shapeless.ops.zipper$Get.get())."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=shapeless.::
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // def typed(t: Function0): Unit in package test, tree.tpe=Unit
[error]         "shapeless"."test"."package"."typed" // def typed(t: Function0): Unit in package test, tree.tpe=(t: Function0)Unit
[error]         Block( // tree.tpe=Function0
[error]           {}
[error]           Typed( // tree.tpe=Function0
[error]             Apply( // def <init>(): shapeless.examples.anonfun$3, tree.tpe=shapeless.examples.anonfun$3
[error]               new anonymous class anonfun$3."<init>" // def <init>(): shapeless.examples.anonfun$3, tree.tpe=()shapeless.examples.anonfun$3
[error]               Nil
[error]             )
[error]             <tpt> // tree.tpe=Function0
[error]           )
[error]         )
[error]       )
[error]       Apply( // def println(x: Object): Unit in object Predef, tree.tpe=Unit
[error]         scala.this."Predef"."println" // def println(x: Object): Unit in object Predef, tree.tpe=(x: Object)Unit
[error]         Apply( // val g2(): shapeless.:: in object ZipperExamples, tree.tpe=shapeless.::
[error]           ZipperExamples$delayedInit$body.this."$outer"."g2" // val g2(): shapeless.:: in object ZipperExamples, tree.tpe=()shapeless.::
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def g3_=(x$1: shapeless.examples.ZipperExamples$Employee): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."g3_$eq" // private[object ZipperExamples] def g3_=(x$1: shapeless.examples.ZipperExamples$Employee): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.examples.ZipperExamples$Employee)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=shapeless.examples.ZipperExamples$Employee
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()shapeless.examples.ZipperExamples$Employee
[error]             ZipperExamples$delayedInit$body.this.$outer.z().get(shapeless.ops.zipper$Get.get())."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=shapeless.examples.ZipperExamples$Employee
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // def typed(t: Function0): Unit in package test, tree.tpe=Unit
[error]         "shapeless"."test"."package"."typed" // def typed(t: Function0): Unit in package test, tree.tpe=(t: Function0)Unit
[error]         Block( // tree.tpe=Function0
[error]           {}
[error]           Typed( // tree.tpe=Function0
[error]             Apply( // def <init>(): shapeless.examples.anonfun$4, tree.tpe=shapeless.examples.anonfun$4
[error]               new anonymous class anonfun$4."<init>" // def <init>(): shapeless.examples.anonfun$4, tree.tpe=()shapeless.examples.anonfun$4
[error]               Nil
[error]             )
[error]             <tpt> // tree.tpe=Function0
[error]           )
[error]         )
[error]       )
[error]       Apply( // def println(x: Object): Unit in object Predef, tree.tpe=Unit
[error]         scala.this."Predef"."println" // def println(x: Object): Unit in object Predef, tree.tpe=(x: Object)Unit
[error]         Apply( // val g3(): shapeless.examples.ZipperExamples$Employee in object ZipperExamples, tree.tpe=shapeless.examples.ZipperExamples$Employee
[error]           ZipperExamples$delayedInit$body.this."$outer"."g3" // val g3(): shapeless.examples.ZipperExamples$Employee in object ZipperExamples, tree.tpe=()shapeless.examples.ZipperExamples$Employee
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def g4_=(x$1: Int): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."g4_$eq" // private[object ZipperExamples] def g4_=(x$1: Int): Unit in object ZipperExamples, tree.tpe=(x$1: Int)Unit
[error]         Apply( // def unbox(x: Object): Int in object Int, tree.tpe=Int
[error]           "scala"."Int"."unbox" // def unbox(x: Object): Int in object Int, tree.tpe=(x: Object)Int
[error]           Apply( // def get(get: shapeless.ops.zipper$Get): Object in class Zipper, tree.tpe=Object
[error]             ZipperExamples$delayedInit$body.this.$outer.z().down(shapeless.ops.zipper$Down.genericDown({
[error]   new shapeless.examples.1()
[error] })).$asInstanceOf[shapeless.Zipper]().right(shapeless.ops.zipper$Right.right()).$asInstanceOf[shapeless.Zipper]()."get" // def get(get: shapeless.ops.zipper$Get): Object in class Zipper, tree.tpe=(get: shapeless.ops.zipper$Get)Object
[error]             Apply( // implicit def get(): shapeless.ops.zipper$Get in object zipper$Get, tree.tpe=shapeless.ops.zipper$Get
[error]               "shapeless"."ops"."zipper$Get"."get" // implicit def get(): shapeless.ops.zipper$Get in object zipper$Get, tree.tpe=()shapeless.ops.zipper$Get
[error]               Nil
[error]             )
[error]           )
[error]         )
[error]       )
[error]       Apply( // def typed(t: Function0): Unit in package test, tree.tpe=Unit
[error]         "shapeless"."test"."package"."typed" // def typed(t: Function0): Unit in package test, tree.tpe=(t: Function0)Unit
[error]         Block( // tree.tpe=Function0
[error]           {}
[error]           Typed( // tree.tpe=Function0
[error]             Apply( // def <init>(): shapeless.examples.anonfun$1, tree.tpe=shapeless.examples.anonfun$1
[error]               new anonymous class anonfun$1."<init>" // def <init>(): shapeless.examples.anonfun$1, tree.tpe=()shapeless.examples.anonfun$1
[error]               Nil
[error]             )
[error]             <tpt> // tree.tpe=Function0
[error]           )
[error]         )
[error]       )
[error]       Apply( // def println(x: Object): Unit in object Predef, tree.tpe=Unit
[error]         scala.this."Predef"."println" // def println(x: Object): Unit in object Predef, tree.tpe=(x: Object)Unit
[error]         Apply( // def box(x: Int): Integer in object Int, tree.tpe=Object
[error]           "scala"."Int"."box" // def box(x: Int): Integer in object Int, tree.tpe=(x: Int)Integer
[error]           Apply( // val g4(): Int in object ZipperExamples, tree.tpe=Int
[error]             ZipperExamples$delayedInit$body.this."$outer"."g4" // val g4(): Int in object ZipperExamples, tree.tpe=()Int
[error]             Nil
[error]           )
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def g5_=(x$1: String): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."g5_$eq" // private[object ZipperExamples] def g5_=(x$1: String): Unit in object ZipperExamples, tree.tpe=(x$1: String)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=String
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()String
[error]             ZipperExamples$delayedInit$body.this.$outer.z().down(shapeless.ops.zipper$Down.genericDown({
[error]   new shapeless.examples.1()
[error] })).$asInstanceOf[shapeless.Zipper]().get(shapeless.ops.zipper$Get.get())."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=String
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // def typed(t: Function0): Unit in package test, tree.tpe=Unit
[error]         "shapeless"."test"."package"."typed" // def typed(t: Function0): Unit in package test, tree.tpe=(t: Function0)Unit
[error]         Block( // tree.tpe=Function0
[error]           {}
[error]           Typed( // tree.tpe=Function0
[error]             Apply( // def <init>(): shapeless.examples.anonfun$5, tree.tpe=shapeless.examples.anonfun$5
[error]               new anonymous class anonfun$5."<init>" // def <init>(): shapeless.examples.anonfun$5, tree.tpe=()shapeless.examples.anonfun$5
[error]               Nil
[error]             )
[error]             <tpt> // tree.tpe=Function0
[error]           )
[error]         )
[error]       )
[error]       Apply( // def println(x: Object): Unit in object Predef, tree.tpe=Unit
[error]         scala.this."Predef"."println" // def println(x: Object): Unit in object Predef, tree.tpe=(x: Object)Unit
[error]         Apply( // val g5(): String in object ZipperExamples, tree.tpe=String
[error]           ZipperExamples$delayedInit$body.this."$outer"."g5" // val g5(): String in object ZipperExamples, tree.tpe=()String
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def g6_=(x$1: shapeless.Zipper): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."g6_$eq" // private[object ZipperExamples] def g6_=(x$1: shapeless.Zipper): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.Zipper)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=shapeless.Zipper
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()shapeless.Zipper
[error]             ZipperExamples$delayedInit$body.this.$outer.z().down(shapeless.ops.zipper$Down.genericDown({
[error]   new shapeless.examples.1()
[error] })).$asInstanceOf[shapeless.Zipper]().put("King Agamemnon", shapeless.ops.zipper$Put.genericPut({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=shapeless.Zipper
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def g8_=(x$1: shapeless.Zipper): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."g8_$eq" // private[object ZipperExamples] def g8_=(x$1: shapeless.Zipper): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.Zipper)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=shapeless.Zipper
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()shapeless.Zipper
[error]             ZipperExamples$delayedInit$body.this.$outer.g6().right(shapeless.ops.zipper$Right.right()).$asInstanceOf[shapeless.Zipper]().put(scala.Int.box(8000), shapeless.ops.zipper$Put.genericPut({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())))."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=shapeless.Zipper
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def g9_=(x$1: shapeless.Zipper): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."g9_$eq" // private[object ZipperExamples] def g9_=(x$1: shapeless.Zipper): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.Zipper)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=shapeless.Zipper
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()shapeless.Zipper
[error]             ZipperExamples$delayedInit$body.this.$outer.g8().up(shapeless.ops.zipper$Up.up(shapeless.ops.zipper$Reify.genericReify({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())), shapeless.ops.zipper$Put.genericPut({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())))."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=shapeless.Zipper
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def agamemnon_=(x$1: shapeless.examples.ZipperExamples$Employee): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."agamemnon_$eq" // private[object ZipperExamples] def agamemnon_=(x$1: shapeless.examples.ZipperExamples$Employee): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.examples.ZipperExamples$Employee)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=shapeless.examples.ZipperExamples$Employee
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()shapeless.examples.ZipperExamples$Employee
[error]             ZipperExamples$delayedInit$body.this.$outer.g9().get(shapeless.ops.zipper$Get.get())."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=shapeless.examples.ZipperExamples$Employee
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // def typed(t: Function0): Unit in package test, tree.tpe=Unit
[error]         "shapeless"."test"."package"."typed" // def typed(t: Function0): Unit in package test, tree.tpe=(t: Function0)Unit
[error]         Block( // tree.tpe=Function0
[error]           {}
[error]           Typed( // tree.tpe=Function0
[error]             Apply( // def <init>(): shapeless.examples.anonfun$6, tree.tpe=shapeless.examples.anonfun$6
[error]               new anonymous class anonfun$6."<init>" // def <init>(): shapeless.examples.anonfun$6, tree.tpe=()shapeless.examples.anonfun$6
[error]               Nil
[error]             )
[error]             <tpt> // tree.tpe=Function0
[error]           )
[error]         )
[error]       )
[error]       Apply( // def println(x: Object): Unit in object Predef, tree.tpe=Unit
[error]         scala.this."Predef"."println" // def println(x: Object): Unit in object Predef, tree.tpe=(x: Object)Unit
[error]         Apply( // val agamemnon(): shapeless.examples.ZipperExamples$Employee in object ZipperExamples, tree.tpe=shapeless.examples.ZipperExamples$Employee
[error]           ZipperExamples$delayedInit$body.this."$outer"."agamemnon" // val agamemnon(): shapeless.examples.ZipperExamples$Employee in object ZipperExamples, tree.tpe=()shapeless.examples.ZipperExamples$Employee
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def updatedDept_=(x$1: shapeless.examples.ZipperExamples$Dept): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."updatedDept_$eq" // private[object ZipperExamples] def updatedDept_=(x$1: shapeless.examples.ZipperExamples$Dept): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.examples.ZipperExamples$Dept)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()shapeless.examples.ZipperExamples$Dept
[error]             ZipperExamples$delayedInit$body.this.$outer.g9().reify(shapeless.ops.zipper$Reify.genericReify({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // def typed(t: Function0): Unit in package test, tree.tpe=Unit
[error]         "shapeless"."test"."package"."typed" // def typed(t: Function0): Unit in package test, tree.tpe=(t: Function0)Unit
[error]         Block( // tree.tpe=Function0
[error]           {}
[error]           Typed( // tree.tpe=Function0
[error]             Apply( // def <init>(): shapeless.examples.anonfun$7, tree.tpe=shapeless.examples.anonfun$7
[error]               new anonymous class anonfun$7."<init>" // def <init>(): shapeless.examples.anonfun$7, tree.tpe=()shapeless.examples.anonfun$7
[error]               Nil
[error]             )
[error]             <tpt> // tree.tpe=Function0
[error]           )
[error]         )
[error]       )
[error]       Apply( // def println(x: Object): Unit in object Predef, tree.tpe=Unit
[error]         scala.this."Predef"."println" // def println(x: Object): Unit in object Predef, tree.tpe=(x: Object)Unit
[error]         Apply( // val updatedDept(): shapeless.examples.ZipperExamples$Dept in object ZipperExamples, tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           ZipperExamples$delayedInit$body.this."$outer"."updatedDept" // val updatedDept(): shapeless.examples.ZipperExamples$Dept in object ZipperExamples, tree.tpe=()shapeless.examples.ZipperExamples$Dept
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def achillesRaise_=(x$1: shapeless.examples.ZipperExamples$Dept): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."achillesRaise_$eq" // private[object ZipperExamples] def achillesRaise_=(x$1: shapeless.examples.ZipperExamples$Dept): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.examples.ZipperExamples$Dept)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()shapeless.examples.ZipperExamples$Dept
[error]             ZipperExamples$delayedInit$body.this.$outer.g9().right(shapeless.ops.zipper$Right.right()).$asInstanceOf[shapeless.Zipper]().down(shapeless.ops.zipper$Down.hlistDown()).$asInstanceOf[shapeless.Zipper]().right(shapeless.ops.zipper$Right.right()).$asInstanceOf[shapeless.Zipper]().down(shapeless.ops.zipper$Down.genericDown({
[error]   new shapeless.examples.1()
[error] })).$asInstanceOf[shapeless.Zipper]().right(shapeless.ops.zipper$Right.right()).$asInstanceOf[shapeless.Zipper]().put(scala.Int.box(3000), shapeless.ops.zipper$Put.genericPut({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))).$asInstanceOf[shapeless.Zipper]().root(shapeless.ops.zipper$Root.nonRootRoot(shapeless.ops.zipper$Up.up(shapeless.ops.zipper$Reify.genericReify({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())), shapeless.ops.zipper$Put.hlistPut(shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))), shapeless.ops.zipper$Root.nonRootRoot(shapeless.ops.zipper$Up.up(shapeless.ops.zipper$Reify.hlistReify(shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())), shapeless.ops.zipper$Put.genericPut({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))), shapeless.ops.zipper$Root.rootRoot()))).$asInstanceOf[shapeless.Zipper]().reify(shapeless.ops.zipper$Reify.genericReify({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())))."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // def typed(t: Function0): Unit in package test, tree.tpe=Unit
[error]         "shapeless"."test"."package"."typed" // def typed(t: Function0): Unit in package test, tree.tpe=(t: Function0)Unit
[error]         Block( // tree.tpe=Function0
[error]           {}
[error]           Typed( // tree.tpe=Function0
[error]             Apply( // def <init>(): shapeless.examples.anonfun$8, tree.tpe=shapeless.examples.anonfun$8
[error]               new anonymous class anonfun$8."<init>" // def <init>(): shapeless.examples.anonfun$8, tree.tpe=()shapeless.examples.anonfun$8
[error]               Nil
[error]             )
[error]             <tpt> // tree.tpe=Function0
[error]           )
[error]         )
[error]       )
[error]       Apply( // def println(x: Object): Unit in object Predef, tree.tpe=Unit
[error]         scala.this."Predef"."println" // def println(x: Object): Unit in object Predef, tree.tpe=(x: Object)Unit
[error]         Apply( // val achillesRaise(): shapeless.examples.ZipperExamples$Dept in object ZipperExamples, tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           ZipperExamples$delayedInit$body.this."$outer"."achillesRaise" // val achillesRaise(): shapeless.examples.ZipperExamples$Dept in object ZipperExamples, tree.tpe=()shapeless.examples.ZipperExamples$Dept
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // private[object ZipperExamples] def singlePass_=(x$1: shapeless.examples.ZipperExamples$Dept): Unit in object ZipperExamples, tree.tpe=Unit
[error]         ZipperExamples$delayedInit$body.this."$outer"."singlePass_$eq" // private[object ZipperExamples] def singlePass_=(x$1: shapeless.examples.ZipperExamples$Dept): Unit in object ZipperExamples, tree.tpe=(x$1: shapeless.examples.ZipperExamples$Dept)Unit
[error]         Apply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           TypeApply( // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=()shapeless.examples.ZipperExamples$Dept
[error]             ZipperExamples$delayedInit$body.this.$outer.z().down(shapeless.ops.zipper$Down.genericDown({
[error]   new shapeless.examples.1()
[error] })).$asInstanceOf[shapeless.Zipper]().put("King Agamemnon", shapeless.ops.zipper$Put.genericPut({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())).$asInstanceOf[shapeless.Zipper]().right(shapeless.ops.zipper$Right.right()).$asInstanceOf[shapeless.Zipper]().put(scala.Int.box(8000), shapeless.ops.zipper$Put.genericPut({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))).$asInstanceOf[shapeless.Zipper]().up(shapeless.ops.zipper$Up.up(shapeless.ops.zipper$Reify.genericReify({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())), shapeless.ops.zipper$Put.genericPut({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))).$asInstanceOf[shapeless.Zipper]().right(shapeless.ops.zipper$Right.right()).$asInstanceOf[shapeless.Zipper]().down(shapeless.ops.zipper$Down.hlistDown()).$asInstanceOf[shapeless.Zipper]().right(shapeless.ops.zipper$Right.right()).$asInstanceOf[shapeless.Zipper]().down(shapeless.ops.zipper$Down.genericDown({
[error]   new shapeless.examples.1()
[error] })).$asInstanceOf[shapeless.Zipper]().right(shapeless.ops.zipper$Right.right()).$asInstanceOf[shapeless.Zipper]().put(scala.Int.box(3000), shapeless.ops.zipper$Put.genericPut({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))).$asInstanceOf[shapeless.Zipper]().root(shapeless.ops.zipper$Root.nonRootRoot(shapeless.ops.zipper$Up.up(shapeless.ops.zipper$Reify.genericReify({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())), shapeless.ops.zipper$Put.hlistPut(shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))), shapeless.ops.zipper$Root.nonRootRoot(shapeless.ops.zipper$Up.up(shapeless.ops.zipper$Reify.hlistReify(shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())), shapeless.ops.zipper$Put.genericPut({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1()))), shapeless.ops.zipper$Root.rootRoot()))).$asInstanceOf[shapeless.Zipper]().reify(shapeless.ops.zipper$Reify.genericReify({
[error]   new shapeless.examples.1()
[error] }, shapeless.ops.hlist$ReversePrepend.hlistReversePrepend(shapeless.ops.hlist$ReversePrepend.hnilReversePrepend1())))."$asInstanceOf" // final def $asInstanceOf[T0 >: ? <: ?](): T0 in class Object, tree.tpe=[T0 >: ? <: ?]()T0
[error]             <tpt> // tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           )
[error]           Nil
[error]         )
[error]       )
[error]       Apply( // def typed(t: Function0): Unit in package test, tree.tpe=Unit
[error]         "shapeless"."test"."package"."typed" // def typed(t: Function0): Unit in package test, tree.tpe=(t: Function0)Unit
[error]         Block( // tree.tpe=Function0
[error]           {}
[error]           Typed( // tree.tpe=Function0
[error]             Apply( // def <init>(): shapeless.examples.anonfun$9, tree.tpe=shapeless.examples.anonfun$9
[error]               new anonymous class anonfun$9."<init>" // def <init>(): shapeless.examples.anonfun$9, tree.tpe=()shapeless.examples.anonfun$9
[error]               Nil
[error]             )
[error]             <tpt> // tree.tpe=Function0
[error]           )
[error]         )
[error]       )
[error]       Apply( // def println(x: Object): Unit in object Predef, tree.tpe=Unit
[error]         scala.this."Predef"."println" // def println(x: Object): Unit in object Predef, tree.tpe=(x: Object)Unit
[error]         Apply( // val singlePass(): shapeless.examples.ZipperExamples$Dept in object ZipperExamples, tree.tpe=shapeless.examples.ZipperExamples$Dept
[error]           ZipperExamples$delayedInit$body.this."$outer"."singlePass" // val singlePass(): shapeless.examples.ZipperExamples$Dept in object ZipperExamples, tree.tpe=()shapeless.examples.ZipperExamples$Dept
[error]           Nil
[error]         )
[error]       )
[error]       "scala"."runtime"."BoxedUnit"."UNIT" // final val UNIT: runtime.BoxedUnit in object BoxedUnit, tree.tpe=runtime.BoxedUnit
[error]     )
[error]   )
[error]   DefDef( // def <init>($outer: shapeless.examples.ZipperExamples.type): shapeless.examples.ZipperExamples$delayedInit$body in class ZipperExamples$delayedInit$body
[error]     <method> <triedcooking>
[error]     "<init>"
[error]     []
[error]     // 1 parameter list
[error]     ValDef( // $outer: shapeless.examples.ZipperExamples.type
[error]       <param> <paramaccessor> <triedcooking>
[error]       "$outer"
[error]       <tpt> // tree.tpe=shapeless.examples.ZipperExamples.type
[error]       <empty>
[error]     )
[error]     <tpt> // tree.tpe=shapeless.examples.ZipperExamples$delayedInit$body
[error]     Block( // tree.tpe=Unit
[error]       // 2 statements
[error]       If( // tree.tpe=Unit
[error]         Apply( // final def eq(x$1: Object): Boolean in class Object, tree.tpe=Boolean
[error]           "$outer"."eq" // final def eq(x$1: Object): Boolean in class Object, tree.tpe=(x$1: Object)Boolean
[error]           null
[error]         )
[error]         Throw( // tree.tpe=Nothing
[error]           Apply( // def <init>(): NullPointerException in class NullPointerException, tree.tpe=NullPointerException
[error]             new NullPointerException."<init>" // def <init>(): NullPointerException in class NullPointerException, tree.tpe=()NullPointerException
[error]             Nil
[error]           )
[error]         )
[error]         Assign( // tree.tpe=Unit
[error]           ZipperExamples$delayedInit$body.this."$outer" // private[this] val $outer: shapeless.examples.ZipperExamples.type in class ZipperExamples$delayedInit$body, tree.tpe=shapeless.examples.ZipperExamples.type
[error]           "$outer" // $outer: shapeless.examples.ZipperExamples.type, tree.tpe=shapeless.examples.ZipperExamples.type
[error]         )
[error]       )
[error]       Apply( // def <init>(): scala.runtime.AbstractFunction0 in class AbstractFunction0, tree.tpe=scala.runtime.AbstractFunction0
[error]         ZipperExamples$delayedInit$body.super."<init>" // def <init>(): scala.runtime.AbstractFunction0 in class AbstractFunction0, tree.tpe=()scala.runtime.AbstractFunction0
[error]         Nil
[error]       )
[error]       ()
[error]     )
[error]   )
[error] )
[error] 
[error] == Expanded type of tree ==
[error] 
[error] TypeRef(TypeSymbol(final abstract class Long extends ))
[error] 
[error] uncaught exception during compilation: java.lang.AssertionError
[trace] Stack trace suppressed: run last examples/compile:compile for the full output.
[error] (examples/compile:compile) java.lang.AssertionError: assertion failed: Unsound substitution from List(type V) to List(type V, value unwrap, method label, method unlabel, type V)
```

--- a/compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala
+++ b/compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala
@@ -831,8 +831,18 @@ abstract class GenJSCode extends plugins.PluginComponent
           }
 
           val call = genApplyMethod(genThis(), sym, jsParams.map(_.ref))
-          val body = ensureBoxed(call,
-              enteringPhase(currentRun.posterasurePhase)(sym.tpe.resultType))
+          val resTpeEnteringPosterasure = enteringPhase(currentRun.posterasurePhase) {
+            sym.tpe match {
+              case _: ExistentialType =>
+                /* We should not see an ExistentialType here. This is a
+                 * scalac 2.10 bug. We assume no boxing is required. See #1581.
+                 */
+                ObjectTpe
+              case symTpe =>
+                symTpe.resultType
+            }
+          }
+          val body = ensureBoxed(call, resTpeEnteringPosterasure)
 
           currentClassInfoBuilder.addMethod(currentMethodInfoBuilder.result())
 

